### PR TITLE
Bluetooth: controller: Extended Scanning with duration and period

### PR DESCRIPTION
**Pull request by dagbja**
Start/stop scan according to period and duration params

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

### CODE REVIEW

**Comment by dagbja** [comment]
hci.c, line 1916,1917 is temporary code to be able to test with shell. Will be removed later.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 1]
```diff
@@ -131,6 +131,16 @@ uint8_t ll_scan_enable(uint8_t enable)
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+	scan->lll.period = (uint64_t)period * (1280000U / 625U);
+	scan->lll.duration = (uint64_t)duration * (10000U / 625U);
+
+	if (scan->lll.duration && scan->lll.period
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:137]: This validation are only required for controller-only build, use conditional compile using CONFIG_BT_CTLR_PARAM_CHECK

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 2]
```diff
@@ -482,6 +492,19 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
 	struct ll_scan_set *scan = param;
 	uint32_t ret;
 	uint8_t ref;
+	struct lll_scan *lll;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:495]: This is incorrect, you have to design a "periodic scanning" for a "duration". You are shunting the "scan interval" after "duration" in your implementation here.

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 2 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:495]: Yes, I'm shunting it until start of the next period, so it will scan for "duration", restarting every "period".

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:495]: By shunting, other BLE states/roles are blocked from using the radio air time as scanner is reserving the window every interval, beats the purpose of period and duration feature.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.h [hunk 3]
```diff
@@ -38,6 +38,9 @@ struct lll_scan {
 
 	uint16_t interval;
 	uint32_t ticks_window;
+	uint32_t period;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.h:41]: move fields to ll_scan_set, as lll_scan member are the ones used/modified by LLL.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 4]
```diff
@@ -480,9 +492,51 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
...
+					(TICKER_ID_SCAN_BASE +
+						ull_scan_handle_get(scan)),
+					0, 0, 0, 0,
+					1,
+					1,
+					NULL, NULL);
+			LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
+				  (ret == TICKER_STATUS_BUSY));
+		} else {
+			lll->dur_counter += lll->interval;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:519]: Use countdown, compared to addition operation.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 5]
```diff
@@ -131,6 +131,18 @@ uint8_t ll_scan_enable(uint8_t enable)
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+	scan->lll.period = (uint64_t)period * (1280000U / 625U);
+	scan->lll.duration = (uint64_t)duration * (10000U / 625U);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:135]: Store in `interval` units and use countdown

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 1 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:138]: Please valid the parameters, before use. Move validation before line 134.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 4 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:519]: Also, need to take `lazy` value into account.

**Review by cvinayak — DISMISSED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 1 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c]: Fixed

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 2 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c]: Modified code to avoid this

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.h (hunk 3 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.h]: Fixed

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 4 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c]: Fixed

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 5 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c]: Fixed

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 1 above)
Comment (dagbja) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c]: Just remembered why it was after. We need to scale period and duration to comare them. Moved the check back after the scaling.

**Comment by cvinayak** [comment]
@Thalley This PR illustrates changes spanning across controller's HCI->ULL->LLL and back to HCI generation of event. Reviewing, debugging its controller path, and testing the implementation here will give understanding of one of the design pattern followed. Most new work would be repetition of these patterns.

@wopu-ot and @kruithofa the PR is a simple implementation of duration and scan period, in units of scan interval, hence the scan period will drift, hope this is acceptable as a initial support? accuracy can be improved by maintaining a `remainder` to nudge the scan period, this should be done in a subsequent PR.

Also, enabling an already enabled scanning needs to update/reset the filter, duration and scan period; I will send follow up commits.

Please do review as appropriate.

File: subsys/bluetooth/controller/ll_sw/lll.h [hunk 6]
```diff
@@ -164,68 +164,70 @@ enum node_rx_type {
...
-	NODE_RX_TYPE_MESH_ADV_CPLT = 0x17,
-	NODE_RX_TYPE_MESH_REPORT = 0x18,
+	NODE_RX_TYPE_MESH_ADV_CPLT = 0x18,
+	NODE_RX_TYPE_MESH_REPORT = 0x19,
 #endif /* CONFIG_BT_HCI_MESH_EXT */
 
 /* Following proprietary defines must be at end of enum range */
 #if defined(CONFIG_BT_CTLR_USER_EXT)
-	NODE_RX_TYPE_USER_START = 0x18,
+	NODE_RX_TYPE_USER_START = 0x1a,
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/lll.h:230]: This value is incremented by 2 but only one type was inserted above. But if I understand correctly, this change actually fixes a lurking issue.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 7]
```diff
@@ -1511,13 +1511,14 @@ void ull_adv_done(struct node_rx_event_done *done)
...
+	LL_ASSERT(handle < BT_CTLR_ADV_SET);
+
 	rx_hdr->type = NODE_RX_TYPE_EXT_ADV_TERMINATE;
+	rx_hdr->handle = handle;
 	rx_hdr->rx_ftr.param_adv_term.conn_handle = 0xffff;
 	rx_hdr->rx_ftr.param_adv_term.num_events = adv->event_counter;
 
-	handle = ull_adv_handle_get(adv);
-	LL_ASSERT(handle < BT_CTLR_ADV_SET);
-
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:1520]: These changes seem to be unrelated to scanning. I would prefer to move them to a separate PR (which then probably will be trivial to review and get merged).

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c [hunk 8]
```diff
@@ -921,6 +929,39 @@ static void test_scanx_main(void)
...
+	scan_param.interval = 0x08;
+	scan_param.timeout = 100;
+	err = bt_le_scan_start(&scan_param, scan_cb);
+	if (err) {
+		goto exit;
+	}
+	printk("success.\n");
+
+	printk("Waiting...");
+	while (!is_scan_timeout) {
```
Comment (wopu-ot) [inline tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:943]: I think it would be better to wait for the set duration and then check the value of is_scan_timeout to avoid hanging if the termination fails.

**Review by wopu-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 9]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+			struct node_rx_pdu *node_rx_scan_term;
+			void *link_scan_term;
+
+			scan->duration_lazy = 0U;
+
+			/* The alloc here used for ext scan termination event */
+			link_scan_term = ll_rx_link_alloc();
+			if (!link_scan_term) {
+				/* TODO: figure out right return value */
+				return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:189]: I would argue that `BT_HCI_ERR_MEM_CAPACITY_EXCEEDED` sounds like a proper return value, and the TODO (both this and the one below) should be removed.

File: subsys/bluetooth/controller/ll_sw/ull_slave.c [hunk 10]
```diff
@@ -230,8 +230,8 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
 		rx = adv->lll.node_rx_adv_term;
 		link = rx->link;
 
-		rx->handle = ull_adv_handle_get(adv);
 		rx->type = NODE_RX_TYPE_EXT_ADV_TERMINATE;
+		rx->handle = ull_adv_handle_get(adv);
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_slave.c:234]: Seems like an unnecessary change.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_slave.c:234]: This is make it consistent, to set the handle after the type.

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c (hunk 8 above)
Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:943]: ok.. I will use k_sleep for 1010 ms and then check if `is_scan_timeout` has been set.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 9 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:189]: What about insufficient resources?

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 11]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+					       625U) / scan->lll.interval;
+			scan->duration_lazy -= lll->duration_reload;
+			scan->node_rx_scan_term = NULL;
+		} else {
+			struct node_rx_pdu *node_rx_scan_term;
+			void *link_scan_term;
+
+			scan->duration_lazy = 0U;
+
+			/* The alloc here used for ext scan termination event */
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:185]: The pool is owned by the thread context, and is to be allocated/released in thread context, hence the reservation.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 12]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period << 7))) {
+				return BT_HCI_ERR_INVALID_PARAM;
+			}
+
+			scan->duration_lazy = ((uint32_t)period * 1280000U /
+					       625U) / scan->lll.interval;
+			scan->duration_lazy -= lll->duration_reload;
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:177]: duration_lazy is directly supplied to ticker_update to stop scan windows from end of duration until scan period ends.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 13]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period << 7))) {
+				return BT_HCI_ERR_INVALID_PARAM;
+			}
+
+			scan->duration_lazy = ((uint32_t)period * 1280000U /
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:175]: Please suggest a define name.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 14]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
...
-#if defined(CONFIG_BT_CTLR_PRIVACY)
-	struct lll_scan *lll;
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	lll = &scan->lll;
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period << 7))) {
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:171]: Suggestion to code change welcome.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 15]
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
 		}
 	}
 
-#if defined(CONFIG_BT_CTLR_PRIVACY)
-	struct lll_scan *lll;
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	lll = &scan->lll;
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:167]: Will be handle as a whole in https://github.com/zephyrproject-rtos/zephyr/issues/23314

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 9 above)
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:189]: That would also be a good alternative. However, it seems that `BT_HCI_ERR_MEM_CAPACITY_EXCEEDED` is typically returned when `ll_rx_xxx_alloc()` fails. See also `ull_adv.c:740`.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (wopu-ot) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:171]: I suggest
`#define EXT_SCAN_PERIOD_UNIT_US 1280000U`
`#define EXT_SCAN_DURATION_UNIT_US 10000U`
...
`... duration >= (uint32_t)period * (EXT_SCAN_PERIOD_UNIT_US / EXT_SCAN_DURATION_UNIT_US) ...`
I am not sure where the best place for the defines would be, but I believe you will find the right place for them.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (wopu-ot) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:177]: I would find it more intuitive to store the period and subtract the duration when updating the ticker, but if the other reviewers are okay with how it is done now, I will not object.

**Comment by cvinayak** [comment]
During testing on nrf52840 target that supported Coded PHY, I discovered issues with terminating scan with two scanning instances. Also, now support for updating scan duration and period of an active scanning instance is added and tested. I will clean up commits and push new commits soon.

Below is the screenshot testing scanning instances:
image

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 16]
```diff
@@ -164,16 +164,22 @@ uint8_t ll_scan_enable(uint8_t enable)
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 	lll = &scan->lll;
 	if (duration) {
-		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
+		lll->duration_reload = ((uint32_t)duration *
+					EXT_SCAN_DURATION_UNIT_US /
+					SCAN_INTERVAL_UNIT_US) /
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c]: I wonder if it would be more readable to have a macro for converting this value, so it'd be `lll->duration_reload = CONVERT(duration, interval)` (with a much better name than `CONVERT` of course. Similar macro can be added for the duration calculation below.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 17]
```diff
@@ -460,6 +480,88 @@ uint8_t ull_scan_disable(uint8_t handle, struct ll_scan_set *scan)
...
+	LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
+		  (ret == TICKER_STATUS_BUSY));
+}
+
+void ull_scan_term_dequeue(uint8_t handle)
+{
+	struct ll_scan_set *scan;
+
+	scan = ull_scan_set_get(handle);
+	scan->is_enabled = 0U;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:537]: Any risk that `ull_scan_set_get` returns NULL here?

Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:537]: Seems like there are plenty of other places where there is a NULL check following this get.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 18]
```diff
@@ -622,41 +728,116 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
 }
 
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
-void ull_scan_done(struct node_rx_event_done *done)
+static uint8_t is_scan_update(uint8_t handle, uint16_t duration,
+			      uint16_t period, struct ll_scan_set **scan,
+			      struct node_rx_pdu **node_rx_scan_term)
+{
+	*scan = ull_scan_set_get(handle);
+	*node_rx_scan_term = (void *)(*scan)->node_rx_scan_term;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:736]: Missing NULL check for `*scan`?

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 19]
```diff
@@ -622,41 +728,116 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
...
+			    (duration >= ((uint32_t)period *
+					  EXT_SCAN_PERIOD_UNIT_US/
+					  EXT_SCAN_DURATION_UNIT_US))) {
+				return BT_HCI_ERR_INVALID_PARAM;
+			}
+
+			scan->duration_lazy = ((uint32_t)period *
+					       EXT_SCAN_PERIOD_UNIT_US /
+					       SCAN_INTERVAL_UNIT_US) /
+					      scan->lll.interval;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:764]: See my other comment regarding conversion macros.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 20]
```diff
@@ -175,7 +232,32 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+	{
+		err = duration_period_update(scan, is_update_1m);
+		if (err) {
+			return err;
+		} else if (is_update_1m) {
+			return 0;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:250]: This is inside an #if defined(CONFIG_BT_CTLR_ADV_EXT) block, no need to check for IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT).

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 21]
```diff
@@ -129,28 +149,65 @@ uint8_t ll_scan_enable(uint8_t enable)
...
+#endif /* CONFIG_BT_CTLR_PHY_CODED */
+	{
+		err = duration_period_setup(scan, duration, period,
+					    &node_rx_scan_term);
+		if (err) {
+			return err;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:200]: This is inside an #if defined(CONFIG_BT_CTLR_ADV_EXT) block, no need to check for IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT).

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 22]
```diff
@@ -510,6 +727,170 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
...
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+static uint8_t is_scan_update(uint8_t handle, uint16_t duration,
+			      uint16_t period, struct ll_scan_set **scan,
+			      struct node_rx_pdu **node_rx_scan_term)
+{
+	*scan = ull_scan_set_get(handle);
+	*node_rx_scan_term = (void *)(*scan)->node_rx_scan_term;
+	return duration && period && (*scan)->lll.duration_reload &&
+	       (*scan)->duration_lazy;
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:738]: I did not find in the spec that old or new duration parameters need to be non-zero for a scan update. But maybe I am overlooking something?

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 23]
```diff
@@ -510,6 +727,170 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
...
+	lll = &scan->lll;
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration *
+					EXT_SCAN_DURATION_UNIT_US /
+					SCAN_INTERVAL_UNIT_US) /
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period *
+					  EXT_SCAN_PERIOD_UNIT_US/
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/ull_scan.c:756]: nitpick: space before /

**Review by wopu-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 22 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:738]: BT Spec v5.2, Vol.4, Part E, Section 7.8.65

> When the Duration and Period parameters are non-zero, the Controller shall scan for the duration of the Duration parameter within a scan period specified by the Period parameter.

and 

> If the HCI_LE_Set_Extended_Scan_Enable command with Enable set to 0x01 is sent while scanning is already enabled, the timers used for duration and period are reset to the new parameter values and a new scan period is started.

For the update to be allowed both duration and period need to be non-zero. As Period parameter specifies 'scan period'.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 17 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:537]: I have only added the check in one place, other places the caller uses a constant valid handle value or `SCAN_HANDLE_1M` or `SCAN_HANDLE_PHY_CODED`

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 18 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:736]: The caller uses a constant valid handle value of `SCAN_HANDLE_1M` or `SCAN_HANDLE_PHY_CODED`

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 22 above)
Comment (wopu-ot) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:740]: Do I understand correctly that you interpret this such that the "new scan period" mentioned in the spec implies that there must be an "old scan period"? The test specification does not seem to test this behavior, btw.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:740]: > you interpret this such that the "new scan period" mentioned in the spec implies that there must be an "old scan period"

Yes.

> The test specification does not seem to test this behavior, btw.

I did not get, there is a test that wants to switch from a scanning with non-zero Duration and zero Period to both non-zero and vice versa? Could you point me to the test cases that I can refer to, to make changes accordingly.

Comment (wopu-ot) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:740]: Looks like I did not express myself clearly... I could not find any test case that uses a non-zero period, and only one that uses a non-zero duration. There does not seem to be any test case that updates these parameters. We therefore cannot use the test spec to infer the intended behavior. I do not find the wording in the BT Spec crystal clear, but I think your interpretation is valid.

**Review by wopu-ot — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 22 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:740]: If the implementation needs to allow updates to scanning irrespective of old duration and period, its is not difficult to fix it based on a requirement/request from users. Additonal implementation is required to release buffers allocated for scan terminate event, if period is updated from zero to non-zero value.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c [hunk 24]
```diff
@@ -562,9 +572,17 @@ static void isr_window(void *param)
...
 	uint32_t ticks_at_start;
 	uint32_t remainder_us;
+	struct lll_scan *lll;
 
 	isr_common_done(param);
 
+	lll = param;
+	if (++lll->chan == 3U) {
+		lll->chan = 0U;
+	}
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:582]: `3U` seems like a magic number in this context. Hard to tell what's going on here. Consider adding a comment about why comparing to 3, and why it's `++lll->chan` rather than `lll->chan`.

**Review by Thalley — CHANGES_REQUESTED** [review]
Small request for added readability, otherwise LGTM.

**Review by Thalley — APPROVED** [review]

**Comment by trond-snekvik** [comment]
This broke the build for non-AE apps: https://buildkite.com/zephyr/zephyr/builds/11604
